ceno_zkvm/src/gadgets/signed.rs

```diff
@@ -0,0 +1,57 @@
+use std::{fmt::Display, mem::MaybeUninit};
+
+use ff_ext::ExtensionField;
+
+use crate::{
+    Value,
+    circuit_builder::CircuitBuilder,
+    error::ZKVMError,
+    expression::Expression,
+    instructions::riscv::constants::{BIT_WIDTH, UInt},
+    witness::LkMultiplicity,
+};
+
+use super::SignedExtendConfig;
+
+/// Interprets a `UInt` value as a 2s-complement signed value.
+///
+/// Uses 1 `WitIn` to represent the MSB of the value.
+pub struct Signed<E: ExtensionField> {
+    pub is_negative: SignedExtendConfig<E>,
+    val: Expression<E>,
+}
+
+impl<E: ExtensionField> Signed<E> {
+    pub fn construct_circuit<NR: Into<String> + Display + Clone, N: FnOnce() -> NR>(
+        cb: &mut CircuitBuilder<E>,
+        name_fn: N,
+        unsigned_val: &UInt<E>,
+    ) -> Result<Self, ZKVMError> {
+        cb.namespace(name_fn, |cb| {
+            let is_negative = unsigned_val.is_negative(cb)?;
+            let val = unsigned_val.value() - (1u64 << BIT_WIDTH) * is_negative.expr();
+
+            Ok(Self { is_negative, val })
+        })
+    }
+
+    pub fn assign_instance(
+        &self,
+        instance: &mut [MaybeUninit<E::BaseField>],
+        lkm: &mut LkMultiplicity,
+        val: &Value<u32>,
```

Comment:
Will it be better to use `i32`? Callers could assign `i32::MAX + 1`, and this could cause unexpeced result If the type is`u32`.